docs/contributing/retries-and-waiters.md

```diff
@@ -0,0 +1,471 @@
+
+_NOTE: The section descibes the current handling with Terraform Plugin SDK resource signatures without `context.Context`. In the future, this codebase will be migrated to the context-aware resource signatures which currently enforce a 20 minute default timeout that conflicts with the timeout with the default `max_retries` value. The Terraform Plugin SDK may be updated to support removing this default 20 minute timeout or the default retry mechanism described here will be updated to prevent context cancellation errors where possible._
+
+### Lower Network Error Retries
+
+Given the very high default number of AWS Go SDK retries and the associated timeout before operators would receive the error, the [`hashicorp/aws-sdk-go-base` codebase](https://github.com/hashicorp/aws-sdk-go-base/blob/57529b4c2d2f8f3b5299d66a829b01259fa800d7/session.go#L108-L130) enforces lower retries (10) for certain network errors that typically cannot be remediated via retries. This roughly equates to 30 seconds of retries.
+
+### Terraform AWS Provider Service Retries
+
+The AWS Go SDK middleware provides hooks for injecting custom logic in the service client handlers. This type of handling is preferred when retry behavior must be applied in many resources, such as an error code that is not marked as automatically retryable by the service API. The Terraform AWS Provider includes some additional retry changing behaviors using this method, which can be found in the `aws/config.go` file. For example:
+
+```go
+client.kafkaconn.Handlers.Retry.PushBack(func(r *request.Request) {
+	if tfawserr.ErrMessageContains(r.Error, kafka.ErrCodeTooManyRequestsException, "Too Many Requests") {
+		r.Retryable = aws.Bool(true)
+	}
+})
+```
+
+## Eventual Consistency
+
+Eventual consistency is a temporary condition where the remote system can return outdated information or errors due to not being strongly read-after-write consistent. This is a pattern found in remote systems that must be highly scaled for broad usage.
+
+Terraform expects any planned resource lifecycle change (create, update, destroy of the resource itself) and planned resource attribute value change to match after being applied. Conversely, operators typically expect that Terraform resources also implement the concept of drift detection for resources and their attributes, which requires reading information back from the remote system after an operation. A common implementation is refreshing the Terraform State information (`d.Set()`) during the `Read` function of a resource after `Create` and `Update`.
+
+These two concepts conflict with each other and require additional handling in Terraform resource logic as shown in the following sections. These issues are _not_ reliably reproducible, especially in the case of writing acceptance testing, so they can be elusive with false positives to verify fixes.
```

Comment:
The caveat here is that Terraform will not report the inconsistency as drift unless the configuration happens to be planned again in the eventual consistency window. Instead, Terraform returns errors because the planned value does not match the applied value.